File: 10/js/1.js
const fs = require('fs');
const { create_record, create_cpu } = require("./CPULib");

(function () {
  const contents = fs.readFileSync("../inputSample2.txt").toString();

  const ops = contents.split("\r\n").filter(x => x.length > 0).map(x => x.split(" "));

  const cpu = create_cpu(),
    rec = create_record(20, 40);
  cpu.setRecord(rec);
  ops.forEach(op => {
    cpu.exec(op[0], op.slice(1));
  });
  console.log(`x = ${cpu.getX()} after ${cpu.getCycle()} cycles`);
  const signals = rec.get().map(([x, cy]) => x * cy);
  console.log("Sum of recorded signals: " + signals.reduce((a, s) => a + s, 0));
})();

File: 10/hs/1.hs
import System.IO
import CPULib

main = do
  h <- openFile "../inputSample2.txt" ReadMode
  contents <- hGetContents h
  let
    ops = map ((\(i:a) -> (i, a)).words) $ lines contents
    cpu = create_cpu 1
    rc = create_record True 20 40
    crt = create_crt False 0 0
    (cpu', rc', _) = cpu_exec_all (cpu, rc, crt) ops
    signals = map (\(x,c) -> x*c) $ record_get rc'
  print $ "x = " ++ (show $ fst cpu') ++ " after " ++ (show $ snd cpu') ++ " cycles."
  print $ "Sum of recorded signals: " ++ (show $ sum signals)
  hClose h